src/rgw/rgw_pubsub_push.cc

```diff
@@ -0,0 +1,309 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "rgw_pubsub_push.h"
+#include <string>
+#include <sstream>
+#include <algorithm>
+#include "include/buffer_fwd.h"
+#include "common/Formatter.h"
+#include "rgw_common.h"
+#include "rgw_data_sync.h"
+#include "rgw_pubsub.h"
+#include "rgw_amqp.h"
+#include <boost/asio/yield.hpp>
+#include <functional>
+#include "rgw_perf_counters.h"
+
+using namespace rgw;
+
+std::string json_format_pubsub_event(const rgw_pubsub_event& event) {
+  std::stringstream ss;
+  JSONFormatter f(false);
+  encode_json("event", event, &f);
+  f.flush(ss);
+  return ss.str();
+}
+
+class RGWPubSubHTTPEndpoint : public RGWPubSubEndpoint {
+private:
+  const std::string endpoint;
+  std::string str_ack_level;
+  typedef unsigned ack_level_t;
+  ack_level_t ack_level; // TODO: not used for now
+  static const ack_level_t ACK_LEVEL_ANY = 0;
+  static const ack_level_t ACK_LEVEL_NON_ERROR = 1;
+
+  // PostCR implements async execution of RGWPostHTTPData via coroutine
+  class PostCR : public RGWPostHTTPData, public RGWSimpleCoroutine {
+  private:
+    RGWDataSyncEnv* const sync_env;
+    bufferlist read_bl;
+    const ack_level_t ack_level;
+
+  public:
+    PostCR(const std::string& _post_data,
```

Comment:
I really don't like the idea of creating new CRs that do almost exactly the same as the ones in `rgw_cr_rest.h`. We should reuse common infrastructure. This also implements stuff differently, so I'm not sure it's 100% correct.

Reply:
This will require refactoring rgw_cr_rest, which, in turn requires writing unit-tests for it.
Overall, I agree that this is a good idea, but would prefer doing that in a parallel  PR.
Current "PushCR" implementation (which ignores the content of the replies), is 40 lines of code, and based on same infrastructure: RGWPostHTTPData and RGWSimpleCoroutine.
To add different ack levels, I would wait for the refactoring, so code for reading replies is not implemented in multiple places.

Reply:
this seems like fair reasoning?